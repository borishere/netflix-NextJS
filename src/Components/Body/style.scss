.body {
  display: flex;
  flex-direction: column;
  padding: 30px 90px;
  background-color: #232323;

  .filters-wrap {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
  }

  .movies-found-label {
    margin-top: 50px;
  }
}