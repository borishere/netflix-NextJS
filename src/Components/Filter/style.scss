.filter-wrap {
  display: flex;
  list-style: none;
  justify-content: space-evenly;
  margin: 0;
  padding: 0;

  .filter-item {
    margin-right: 25px;
    cursor: pointer;

    &.active {
      text-decoration: underline;
      text-decoration-color: red;
    }
  }
}
