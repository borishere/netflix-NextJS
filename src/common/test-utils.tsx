import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ModalContext } from '../Context/ModalContext';
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import { createMemoryHistory } from 'history';
import { moviesApi } from '../services/movies';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import moviesReducer from '../app/moviesSlice';
import { configureStore } from '@reduxjs/toolkit';

const history = createMemoryHistory();


const AllTheProviders: FC = ({ children }) => {
  return (
    <Router location='/' navigator={history}>
      <Provider store={store}>
        <ModalContext.Provider value={{}}>
          {children}
        </ModalContext.Provider>
      </Provider>
    </Router>
  );
};

function renderWithStore(
  ui: ReactElement,
  {
    store = configureStore({
      reducer: {
        [moviesApi.reducerPath]: moviesApi.reducer,
        movies: moviesReducer

      },
      middleware: (getDefaultMiddleware: any) =>
        getDefaultMiddleware().concat(moviesApi.middleware)
    }),
    ...renderOptions
  } = {}
) {
  setupListeners(store.dispatch);

  function Wrapper({ children }: any) {
    const history = createMemoryHistory();

    return (
      <Provider store={store}>
        <Router location='/' navigator={history}>
          {children}
        </Router>
      </Provider>
    );
  }

  store.dispatch(moviesApi.util.resetApiState());

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

const root = document.createElement('div');
root.id = 'root';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, container: document.body.appendChild(root), ...options });

export * from '@testing-library/react';
export { customRender, renderWithStore };