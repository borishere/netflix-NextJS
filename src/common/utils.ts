export function addParamToExistsSearchParams(key: string, value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.set(key, value);

  return params;
}

export function deleteParamFromExistsSearchParams(value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.delete(value);

  return params;
}

export const BASE_API_URL: string = 'http://localhost:4000/';