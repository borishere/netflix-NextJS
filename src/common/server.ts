import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { BASE_API_URL } from './utils';
import mockData from './__mocks__/movies.json';

let movies: any[] = Array.from(mockData);

export const server = setupServer(
  rest.get(`${BASE_API_URL}movies`, (req, res, ctx) => {
    return res(ctx.json(movies.slice(0, 2)));
  }),

  rest.get(`${BASE_API_URL}movies:id`, (req, res, ctx) => {
    const { id } = req.params;
    const movie = movies.find((movie) => movie.id === id);

    return res(ctx.json(movie));
  }),

  rest.post(`${BASE_API_URL}movies`, (req, res, ctx) => {
    movies.push(req.body);

    return res(ctx.status(200));
  }),

  rest.put(`${BASE_API_URL}movies`, (req, res, ctx) => {
    const requestMovie: any = req.body;

    const movieIndex = movies.findIndex((m) => m.id === requestMovie.id);

    if (movieIndex < 0) {
      return res(ctx.status(404));
    }

    movies[movieIndex] = requestMovie;

    return res(ctx.status(200));
  }),

  rest.delete(`${BASE_API_URL}movies/:id`, (req, res, ctx) => {
    const { id } = req.params;

    movies = movies.filter((m) => m.id !== id);

    return res(ctx.status(200));
  })
);
