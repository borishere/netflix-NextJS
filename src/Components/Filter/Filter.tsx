import React, { FC, useCallback, useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useAppDispatch } from '../../Hooks/hooks';
import { Igenre } from '../../Models/models';
import { setFilter } from '../../app/moviesSlice';
import { useSearchParams } from 'react-router-dom';
import { addParamToExistsSearchParams, deleteParamFromExistsSearchParams } from '../../common/utils';
import './style.scss';

interface Props {
  genresList: Igenre[];
}

export const Filter: FC<Props> = ({ genresList }) => {
  const [genres, setGenres] = useState<Igenre[]>(genresList);

  const [searchParams] = useSearchParams();
  const genre = searchParams.get('genre');

  const dispatch = useAppDispatch();
  const navigate = useNavigate();

  const genreClickHandler = useCallback((name: string): void => {
    const updatedGenres = genres.map((genre) => {
      genre.active = genre.name === name;

      return genre;
    });

    setGenres([...updatedGenres]);

    let params: URLSearchParams;

    if (name === 'all') {
      params = deleteParamFromExistsSearchParams('genre');
    } else {
      params = addParamToExistsSearchParams('genre', name);
    }

    navigate({ search: params.toString() });
  }, [navigate]);

  useEffect(() => {
    if (genre) {
      genreClickHandler(genre);
    }
  }, [genre, genreClickHandler]);

  useEffect(() => {
    const preparedGenres = genres
      .filter((genre) => genre.active)
      .map((genre) => genre.name);

    dispatch(setFilter(preparedGenres));
  }, [genres, dispatch]);

  return (
    <ul className='filter-wrap'>
      {genres.map((genre) => (
        <li
          key={genre.name}
          className={`filter-item ${genre.active ? 'active' : ''}`}
          onClick={() => genreClickHandler(genre.name)}
        >
          {genre.name.toUpperCase()}
        </li>
      ))}
    </ul>
  );
};
