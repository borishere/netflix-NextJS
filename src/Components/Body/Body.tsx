import React, { FC } from 'react';
import { useCachedMovies } from '../../Hooks/hooks';
import { Igenre } from '../../Models/models';
import { Filter } from '../Filter/Filter';
import { MoviesList } from '../MoviesList/MoviesList';
import { Sorting } from '../Sorting/Sorting';
import './style.scss';

export const genresList: Igenre[] = [
  { name: 'all', active: true },
  { name: 'documentary', active: false },
  { name: 'comedy', active: false },
  { name: 'horror', active: false },
  { name: 'crime', active: false }
];

interface Props {
}

export const Body: FC<Props> = () => {
  const { movies } = useCachedMovies();

  return (
    <div className='body'>
      <div className='filters-wrap'>
        <Filter genresList={genresList} />
        <Sorting />
      </div>
      <div className='movies-found-label'>
        {`${movies?.length} movies found`}
      </div>
      <MoviesList />
    </div>
  );
};