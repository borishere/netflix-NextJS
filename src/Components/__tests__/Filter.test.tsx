import React from 'react';
import { screen, customRender } from '../../common/test-utils';
import { Filter } from '../Filter/Filter';
import { genresList } from '../Body/Body';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

describe('Filter', () => {
  test('Renders correctly', () => {
    const { asFragment } = customRender(<Filter genresList={genresList} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('Genres present', () => {
    customRender(<Filter genresList={genresList} />);

    expect(screen.getByText(/all/i)).toBeInTheDocument();
    expect(screen.getByText(/documentary/i)).toBeInTheDocument();
    expect(screen.getByText(/comedy/i)).toBeInTheDocument();
    expect(screen.getByText(/horror/i)).toBeInTheDocument();
    expect(screen.getByText(/crime/i)).toBeInTheDocument();
  });

  test('Changes classname on click', () => {
    customRender(<Filter genresList={genresList} />);

    const allGenre = screen.getByText(/all/i);
    const documentaryGenre = screen.getByText(/documentary/i);
    const comedyGenre = screen.getByText(/comedy/i);
    const horrorGenre = screen.getByText(/horror/i);
    const crimeGenre = screen.getByText(/crime/i);

    userEvent.click(comedyGenre);

    expect(comedyGenre).toHaveClass('active');
    expect(allGenre).not.toHaveClass('active');
    expect(documentaryGenre).not.toHaveClass('active');
    expect(horrorGenre).not.toHaveClass('active');
    expect(crimeGenre).not.toHaveClass('active');
  });
});
