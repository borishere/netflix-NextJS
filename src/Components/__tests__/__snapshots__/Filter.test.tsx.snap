// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Renders correctly 1`] = `
<DocumentFragment>
  <ul
    class="filter-wrap"
  >
    <li
      class="filter-item active"
    >
      ALL
    </li>
    <li
      class="filter-item "
    >
      DOCUMENTARY
    </li>
    <li
      class="filter-item "
    >
      COMEDY
    </li>
    <li
      class="filter-item "
    >
      HORROR
    </li>
    <li
      class="filter-item "
    >
      CRIME
    </li>
  </ul>
</DocumentFragment>
`;
